### PR TITLE
Support named config binding for remaining Aspire resources

### LINKED ISSUE #5826: Support named configuration providers for Aspire resources
Aspire resource support `Configuration`-based models for configuring resources under well-defined keys. For example, Storage Blob resources support configuration under the `Aspire:Azure:Storage:Blobs` key.

```
{
  "Aspire": {
    "Azure": {
      "Storage": {
        "Blobs": {
          "DisableHealthChecks": true,
          "DisableTracing": false,
          "ClientOptions": {
            "Diagnostics": {
              "ApplicationId": "myapp"
            }
          }
        }
      }
    }
  }
}
```

This configuration model doesn't currently support different configurations for different instances of the same resource type in an application. To address this, we'll update the resource configuration strategy to support instances keyed by name as follows:

```
{
  "Aspire": {
    "Azure": {
      "Storage": {
        "Blobs": {
           "blob1": { .. },
           "blob2": { ..},
        }
      }
    }
  }
}
```

At the moment, the primary motivation for doing this is to provide a resolution for the "dual config dilemma" documented in https://github.com/dotnet/aspire/issues/5563 since the existing configuration provider key name scheme doesn't conflict with Azure WebJob's lookup semantics.

**Tracked Integrations**

- [x] Aspire.Azure.AI.OpenAI
- [x] Aspire.Azure.Data.Tables
- [x] Aspire.Azure.Messaging.EventHubs
- [x] Aspire.Azure.Messaging.ServiceBus
- [x] Aspire.Azure.Message.WebPubSub
- [x] Aspire.Azure.Search.Documents
- [x] Aspire.Azure.Security.KeyVault
- [x] Aspire.Azure.Storage.Blobs
- [x] Aspire.Azure.Storage.Queues
- [x] Aspire.Confluent.Kafka
- [x] Aspire.Elastic.Clients.Elasticsearch
- [x] Aspire.Microsoft.Azure.Cosmos
- [x] Aspire.Microsoft.Data.SqlClient
- [x] Aspire.Microsoft.EntityFrameworkCore.Cosmos
- [x] Aspire.Microsoft.EntityFrameworkCore.SqlServer
- [x] Aspire.Milvus.Client
- [x] Aspire.MongoDB.Driver
- [x] Aspire.MySqlConnector
- [x] Aspire.NATS.Net
- [x] Aspire.Npgsql
- [x] Aspire.Npgsql.EntityFrameworkCore.PostgreSQL
- [ ] Aspire.OpenAI
- [x] Aspire.Oracle.EntityFrameworkCore
- [x] Aspire.Pomelo.EntityFrameworkCore.MySql
- [x] Aspire.Qdrant.Client
- [x] Aspire.RabbitMQ.Client
- [x] Aspire.Seq
- [x] Aspire.StackExchange.Redis
- [x] Aspire.StackExchange.Redis.DistributedCaching
- [x] Aspire.StackExchange.Redis.OutputCaching

### PR DESCRIPTION
**Pull request by captainsafia**
## Description

Update remaining Aspire resources to support binding from top-level and named config.

Fixes https://github.com/dotnet/aspire/issues/5826

## Checklist

- Is this feature complete?
  - [X] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [X] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [X] No
- Does the change require an update in our Aspire docs?
  - [X] Yes
    - Link to aspire-docs issue: https://github.com/dotnet/docs-aspire/issues/1663
  - [ ] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6045)

### CODE REVIEW

**Review by sebastienros — APPROVED** [review]

**Comment by sebastienros** [comment]
Lots of tests are failing though.

Have you checked the OpenAI modules? Asking for a friend.

**Comment by captainsafia** [comment]
> Lots of tests are failing though. (Update: just saw it was a draft PR, my bad)

Yep, running the full test suite is super slow on my machine so relying on CI to help me catch things I'm not covering in the targeted local runs.

> Have you checked the OpenAI modules? Asking for a friend.

Yep, the `Azure.OpenAI` package is already covered by the work in my previous PR (https://github.com/dotnet/aspire/pull/5827). 

I changed the binding logic for the `Aspire.OpenAI` in this PR. I think those are the only two OpenAI things?

**Comment by captainsafia** [comment]
Build is green and ready for review now.

cc: @eerhardt

File: src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/AspireAzureEFCoreCosmosExtensions.cs [hunk 1]
```diff
@@ -48,7 +48,12 @@ public static class AspireAzureEFCoreCosmosExtensions
 
         var settings = builder.GetDbContextSettings<TContext, EntityFrameworkCoreCosmosSettings>(
             DefaultConfigSectionName,
-            (settings, section) => section.Bind(settings)
+            (settings, section) =>
+            {
+                var namedConfigSection = section.GetSection(connectionName);
+                section.Bind(settings);
+                namedConfigSection.Bind(settings);
```
Comment (eerhardt) [inline src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/AspireAzureEFCoreCosmosExtensions.cs:55]: I don't think we need to bind by `connectionName` in these EF libraries. That would make 4 places we look:

* Default
* Default:`typeof(TContext).Name`
* Default:`connectionName`
* Default:`typeof(TContext).Name`:`connectionName`


I think we only need to do the first 2.

File: src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs [hunk 2]
```diff
@@ -55,35 +55,35 @@ public static void AddKafkaConsumer<TKey, TValue>(this IHostApplicationBuilder b
     public static void AddKeyedKafkaConsumer<TKey, TValue>(this IHostApplicationBuilder builder, string name)
```
Comment (eerhardt) [inline src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:55]: Can you also add a blurb to https://github.com/dotnet/aspire/blob/main/src/Components/README.md#configuration about this new pattern?

File: tests/Aspire.Azure.AI.OpenAI.Tests/ConformanceTests.cs [hunk 3]
```diff
@@ -25,6 +25,9 @@ public class ConformanceTests : ConformanceTests<AzureOpenAIClient, AzureOpenAIS
 
     protected override bool SupportsKeyedRegistrations => true;
 
+    protected override bool SupportsNamedConfig => true;
```
Comment (eerhardt) [inline tests/Aspire.Azure.AI.OpenAI.Tests/ConformanceTests.cs:28]: How do you feel about flipping this in the base? Basically the base says `virtual bool SupportsNamedConfig => true;` and then only the places that don't override and return false.

**Review by eerhardt — COMMENTED** [review]
Nice work - I have a couple comments. I think once those are resolved this will be ready to merge.

File: tests/Aspire.Azure.AI.OpenAI.Tests/ConformanceTests.cs (hunk 3 above)
Comment (captainsafia) [reply tests/Aspire.Azure.AI.OpenAI.Tests/ConformanceTests.cs:28]: Yeah, it made sense to have this false-by-default before I add support for everything but flipped it does make sense now. 😄

File: src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs [hunk 4]
```diff
@@ -209,11 +209,14 @@ void OnStatistics(IConsumer<TKey, TValue> _, string json)
     private static KafkaConsumerSettings BuildConsumerSettings(IHostApplicationBuilder builder, string configurationSectionName, Action<KafkaConsumerSettings>? configureSettings, string connectionName)
```
[the remark this answers is not in the document]
Comment (captainsafia) [reply src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs]: OK -- I think I cleaned things up sufficiently here. Let me know what you think.

File: src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs (hunk 2 above)
Comment (captainsafia) [reply src/Components/Aspire.Confluent.Kafka/AspireKafkaConsumerExtensions.cs:55]: I added:

> - Each component should support binding arguments from the configuration section and from named configuration as defined below. Settings bound from named configuration should override those bound from integration-specific configuration.

File: src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/AspireAzureEFCoreCosmosExtensions.cs (hunk 1 above)
Comment (captainsafia) [reply src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/AspireAzureEFCoreCosmosExtensions.cs]: Ah yes, I believe you clarified this for `EnrichDbContext` but I hadn't extended the same philosophy to the `AddDbContext` methods.

File: src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs [hunk 5]
```diff
@@ -69,9 +69,11 @@ private static void AddOpenAIClient(
         ArgumentNullException.ThrowIfNull(builder);
 
         var configSection = builder.Configuration.GetSection(configurationSectionName);
+        var namedConfigSection = configSection.GetSection(connectionName);
 
         OpenAISettings settings = new();
         configSection.Bind(settings);
+        namedConfigSection.Bind(settings);
```
Comment (eerhardt) [inline src/Components/Aspire.OpenAI/AspireOpenAIExtensions.cs:76]: We should update line 91 as well below to bind the ClientOptions to the named config section.

**Review by eerhardt — APPROVED** [review]
This LGTM once the OpenAI comment above is resolved.

Thanks for the contribution.

**Pull request merged**